examples/nlp/text_generation_with_miniature_gpt.py

```diff
@@ -0,0 +1,355 @@
+    return x, y
+
+text_ds = text_ds.map(prepare_lm_inputs_labels)
+text_ds = text_ds.repeat().prefetch(tf.data.experimental.AUTOTUNE)
+
+
+"""
+## Callback for generating text
+"""
+
+
+class TextGenerator(keras.callbacks.Callback):
+    """
+    Callback to generate text from trained model.
+    1. Feed some starting prompt to the model
+    2. Predict probabilities for next token
+    3. Sample next token and add it to the next input
+    
+    # Arguments
+        max_tokens: Integer, the number of tokens to be generated after prompt.
+        start_tokens: List of integers, the token indices for the starting prompt.
+        index_to_word: List of strings, obatined from TextVectorization layer.
+        temp: Float, dictates the variation during sampling.
+            high temp = less probable tokens are also likely to be sampled
+            low temp = highly probable tokens get picked up
+        print_every: Integer, print after this many epochs.
+    """
+
+    def __init__(self, max_tokens, start_tokens, index_to_word, temp=1.0, print_every=5):
+        self.max_tokens = max_tokens
+        self.start_tokens = start_tokens
+        self.index_to_word = index_to_word
+        self.print_every = print_every
+        self.temp = temp
+
+    def sample_from(self, preds):
+        preds = np.asarray(preds).astype("float64")
+        preds = np.log(preds) / self.temp
+        exp_preds = np.exp(preds)
+        preds = exp_preds / np.sum(exp_preds)
+        probas = np.random.multinomial(1, preds, 1)
+        return np.argmax(probas)
+
+    def detokenize(self, number):
+        if number == 0:
+            return '[PAD]'
+        if number == 1:
+            return '[OOV]'
+        return self.index_to_word[number-2].decode("utf-8") 
+
+    def on_epoch_end(self, epoch, logs=None):
+        if (epoch + 1) % self.print_every != 0:
+            return
+        print(f'\n\nepoch = {epoch+1}')
+        num_tokens_generated = 0
+        tokens_generated = []
+        while num_tokens_generated <= self.max_tokens:
+            pad_len = maxlen - len(self.start_tokens)
+            if pad_len < 0:
+                x = self.start_tokens[:maxlen]
+            elif pad_len > 0:
+                x = self.start_tokens + [0] * pad_len
+            else:
+                x = self.start_tokens
+            x = np.array([x])
+            assert x.shape == (1,200), x.shape
+            y,_ = self.model.predict(x)
+            sample_token = self.sample_from(y[0][len(self.start_tokens)-1])
+            tokens_generated.append(sample_token)
+            self.start_tokens.append(sample_token)
+            num_tokens_generated = len(tokens_generated)
+        txt = ' '.join([self.detokenize(_) for _ in self.start_tokens + tokens_generated])
+        print(f'{txt}\n\n')
+
+
+# Tokenize starting prompt
+word_to_index = {}
+for index, word in enumerate(vocab):
+    word_to_index[word.decode('utf-8')] = index + 2  # PAD and OOV at index 0 and 1
```

Comment:
We recently changed the `TextVectorization.get_vocabulary` API (it's an experimental API, so subject to improvements). It returns strings (so no need to decode) and it already includes padding and oov (so no need for the +2 offset).

Reply:
Yeah in one particular session, I was getting strings + pad and OOV tokens in the vocabulary. And then I restarted the session and got bytes 😅

Reply:
Please use a recent tf-nightly version (so the example will be future-proof). It would be good to also add a comment to indicate that the example should be run with tf-nightly higher than [FILL ME] or with TF 2.3 or higher.

Reply:
Yes. Reinstalled tf-nightly (`2.3.0-dev20200531`) and everything is working smoothly except TPUs. The GPU is running pretty fast, so this example is very much doable without TPU now.

Reply:
The TPU error has been reported before ([31st march](https://github.com/tensorflow/tensorflow/issues/38069)). And apparently it was fixed in some versions but has reappeared now.

Reply:
With the stable release and some older versions of tf-nightly, this issue does not come up, but then the session crashes unexpectedly during model.fit